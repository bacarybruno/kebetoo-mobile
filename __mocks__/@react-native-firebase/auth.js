import authors from '@fixtures/authors';

const user = {
  ...authors[0],
  updateProfile: jest.fn(),
};

export const authCredentials = {
  user,
  additionalUserInfo: {
    profile: { id: 'mock-user-profile-id' },
  },
};

const mockOptions = {
  createUserAndRetrieveDataWithEmailAndPassword: jest.fn().mockReturnValue(Promise.resolve(true)),
  sendPasswordResetEmail: jest.fn().mockReturnValue(Promise.resolve(true)),
  signInAndRetrieveDataWithEmailAndPassword: jest.fn().mockReturnValue(Promise.resolve(true)),
  fetchSignInMethodsForEmail: jest.fn().mockReturnValue(Promise.resolve(true)),
  signOut: jest.fn().mockReturnValue(Promise.resolve(true)),
  onAuthStateChanged: jest.fn(),
  signInWithCredential: jest.fn().mockReturnValue(authCredentials),
  credentials: authCredentials,
  signInWithEmailAndPassword: jest.fn().mockReturnValue(Promise.resolve({ user })),
  createUserWithEmailAndPassword: jest.fn().mockReturnValue(Promise.resolve({ user })),
  currentUser: user,
};

const auth = jest.fn(() => mockOptions);

const socialAuthProviderData = {
  credential: jest.fn().mockReturnValue({
    providerId: 'jest',
    secret: 'mock-google-credentials-secret',
    token: 'mock-google-credentials-token',
  }),
};

auth.GoogleAuthProvider = socialAuthProviderData;

auth.FacebookAuthProvider = socialAuthProviderData;

auth.setMockOptions = (options) => {
  Object.keys(options).forEach((key) => {
    mockOptions[key] = options[key];
  });
};

export default auth;
