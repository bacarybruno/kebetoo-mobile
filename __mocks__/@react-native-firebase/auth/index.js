const mockOptions = {
  createUserAndRetrieveDataWithEmailAndPassword: jest.fn().mockReturnValue(Promise.resolve(true)),
  sendPasswordResetEmail: jest.fn().mockReturnValue(Promise.resolve(true)),
  signInAndRetrieveDataWithEmailAndPassword: jest.fn().mockReturnValue(Promise.resolve(true)),
  fetchSignInMethodsForEmail: jest.fn().mockReturnValue(Promise.resolve(true)),
  signOut: jest.fn().mockReturnValue(Promise.resolve(true)),
  onAuthStateChanged: jest.fn(),
  currentUser: null,
}

const auth = jest.fn(() => mockOptions)

auth.setMockOptions = (options) => {
  Object.keys(options).forEach((key) => {
    mockOptions[key] = options[key]
  })
}

export default auth
