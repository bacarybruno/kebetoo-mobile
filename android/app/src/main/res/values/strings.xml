<resources>
    <string name="app_name">Kebetoo</string>
</resources>
