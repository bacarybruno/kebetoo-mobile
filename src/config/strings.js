import LocalizedStrings from 'react-native-localization';
import dayjs from 'dayjs';
import relativeTime from 'dayjs/plugin/relativeTime';
import 'dayjs/locale/en';
import 'dayjs/locale/fr';

import frStrings from '@assets/locales/fr';
import enStrings from '@assets/locales/en';

const strings = new LocalizedStrings({
  fr: frStrings,
  en: enStrings,
});

dayjs.locale(strings.getLanguage());
dayjs.extend(relativeTime);

export const updateAppLocale = (locale) => {
  dayjs.locale(locale);
  strings.setLanguage(locale);
};

export default strings;
