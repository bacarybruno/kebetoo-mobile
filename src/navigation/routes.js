const routes = {
  ONBARDING_NAV: 'route:nav:onboarding',
  SIGNIN: 'route:signin',
  SIGNUP: 'route:signup',
  ONBOARDING: 'route:obnoarding',
  HOME_NAV: 'route:nav:home',
  HOME: 'route:home',
  PROFILE: 'route:profile',
  SEARCH: 'route:search',
  STORIES: 'route:stories',
  CREATE_POST: 'route:create_post',
  TABS_FAB: 'route:custom:tabs_fab',
}

export default routes
