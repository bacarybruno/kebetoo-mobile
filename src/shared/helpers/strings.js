export const capitalize = (s) => {
  if (typeof s !== 'string') return ''
  const splitCharacters = '. '
  const parts = s.split(splitCharacters)
  return parts
    .map((s) => s.charAt(0).toUpperCase() + s.slice(1).toLowerCase())
    .join(splitCharacters)
}
