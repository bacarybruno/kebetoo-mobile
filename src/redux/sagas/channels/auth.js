import { take, call, put } from 'redux-saga/effects'
import { eventChannel } from 'redux-saga'
import auth from '@react-native-firebase/auth'

import * as types from '../../types'

function getAuthChannel() {
  if (!this.authChannel) {
    this.authChannel = eventChannel((emit) => {
      const unsubscribe = auth().onUserChanged((user) => {
        let payload = { isLoggedIn: false }

        if (user) {
          payload = {
            ...payload,
            uid: user.uid,
            isLoggedIn: true,
            email: user.email,
            photoURL: user.photoURL,
            displayName: user.displayName,
            providerUid: user.providerData[0].uid,
            provider: user.providerData[0].providerId,
          }
        }

        if (payload.provider === 'password' && !payload.displayName) {
          // user have been created with email and password
          // but displayName is not yet assigned
          // we wait for the next event that will be assigning displayName to user
          return
        }

        emit({ payload })
      })
      return unsubscribe
    })
  }
  return this.authChannel
}

function* watchForFirebaseAuth() {
  const authChannel = yield call(getAuthChannel)
  try {
    while (true) {
      const { payload } = yield take(authChannel)
      console.log('Received auth event', payload)
      yield put({ type: types.SET_USER_PROFILE_REQUEST, payload })
    }
  } catch (error) {
    console.log('An error occured in auth channel', error)
  } finally {
    console.log('Auth channel terminated gracefully')
  }
}

export default watchForFirebaseAuth
