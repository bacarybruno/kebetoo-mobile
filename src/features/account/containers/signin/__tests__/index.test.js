import { render, fireEvent } from 'react-native-testing-library'
import auth from '@react-native-firebase/auth'
import { act } from 'react-test-renderer'

import setupTest from '@app/config/jest-setup'
import strings from '@app/config/strings'
import routes from '@app/navigation/routes'

import SignIn, { fieldNames } from '../index'

afterEach(jest.clearAllMocks)

const userInfos = {
  email: 'bacarybruno@kebetoo.com',
  password: 'Kebetoo!',
}

const givenSignIn = setupTest(SignIn, render)({
  navigation: {
    setOptions: jest.fn(),
    navigate: jest.fn(),
  },
})

it('renders Signin', () => {
  const { wrapper } = givenSignIn()
  expect(wrapper.toJSON()).toMatchSnapshot()
})

test('screen title', () => {
  const { props } = givenSignIn()
  expect(props.navigation.setOptions).toBeCalledTimes(1)
  expect(props.navigation.setOptions).toBeCalledWith({ title: strings.auth.signin })
})

it('navigates to signup', () => {
  const { wrapper, props } = givenSignIn()
  fireEvent.press(wrapper.getByText(strings.auth.signup))
  expect(props.navigation.navigate).toBeCalledTimes(1)
  expect(props.navigation.navigate).toBeCalledWith(routes.SIGNUP)
})

describe('submit', () => {
  const getElements = (wrapper) => ({
    emailInput: wrapper.getByPlaceholder(strings.auth.email),
    passwordInput: wrapper.getByPlaceholder(strings.auth.password),
    signInButton: wrapper.getByText(strings.auth.signin.toUpperCase()),
  })

  it('throws error if infos are not valid', async () => {
    const { wrapper } = givenSignIn()
    const { signInWithEmailAndPassword } = auth()
    const { emailInput, passwordInput, signInButton } = getElements(wrapper)
    fireEvent.changeText(emailInput, 'fake-mail')
    fireEvent(emailInput, 'onSubmitEditing')
    fireEvent.changeText(passwordInput, '123')
    await fireEvent.press(signInButton)
    expect(signInWithEmailAndPassword).toBeCalledTimes(0)
  })
  it('connects the user if infos are valid', async () => {
    const { wrapper } = givenSignIn()
    const { signInWithEmailAndPassword } = auth()
    const { emailInput, passwordInput, signInButton } = getElements(wrapper)
    fireEvent.changeText(emailInput, userInfos.email)
    fireEvent(emailInput, 'onSubmitEditing')
    fireEvent.changeText(passwordInput, userInfos.password)
    await fireEvent.press(signInButton)
    expect(signInWithEmailAndPassword).toBeCalledTimes(1)
    expect(signInWithEmailAndPassword).toBeCalledWith(userInfos.email, userInfos.password)
  })
})

describe('form fields validation', () => {
  let givenSignInTestRenderer
  beforeEach(() => {
    givenSignInTestRenderer = setupTest(SignIn)({
      navigation: {
        setOptions: jest.fn(),
        navigate: jest.fn(),
      },
    })
  })
  it('is invalid', () => {
    const { wrapper } = givenSignInTestRenderer()
    const emailInput = wrapper.root.findByProps({ fieldName: fieldNames.email })
    act(() => {
      emailInput.props.onValueChange('invalidemail', fieldNames.email)
    })
    act(() => {
      emailInput.props.onBlur(fieldNames.email)
    })
    expect(emailInput.props.error).toBeTruthy()
  })
  it('is valid', () => {
    const { wrapper } = givenSignInTestRenderer()
    const emailInput = wrapper.root.findByProps({ fieldName: fieldNames.email })
    act(() => {
      emailInput.props.onValueChange('bacarybruno@gmail.com', fieldNames.email)
    })
    act(() => {
      emailInput.props.onBlur(fieldNames.email)
    })
    expect(emailInput.props.error).toBeFalsy()
  })
})
