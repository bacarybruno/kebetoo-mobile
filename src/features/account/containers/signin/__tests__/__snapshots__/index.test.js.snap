// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Signin 1`] = `
<RCTScrollView
  automaticallyAdjustContentInsets={false}
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  contentInset={
    Object {
      "bottom": 0,
    }
  }
  enableAutomaticScroll={true}
  enableOnAndroid={true}
  enableResetScrollToCoords={true}
  extraHeight={75}
  extraScrollHeight={24}
  getScrollResponder={[Function]}
  handleOnScroll={[Function]}
  keyboardDismissMode="interactive"
  keyboardOpeningTime={250}
  keyboardShouldPersistTaps="handled"
  keyboardSpace={0}
  onScroll={[Function]}
  resetKeyboardSpace={[Function]}
  scrollEventThrottle={1}
  scrollForExtraHeightOnAndroid={[Function]}
  scrollIntoView={[Function]}
  scrollToEnd={[Function]}
  scrollToFocusedInput={[Function]}
  scrollToPosition={[Function]}
  showsVerticalScrollIndicator={true}
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 255)",
    }
  }
  update={[Function]}
  viewIsInsideTabBar={false}
>
  <View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(255, 255, 255, 255)",
          "flex": 1,
          "paddingHorizontal": 16,
          "paddingVertical": 24,
        }
      }
    >
      <View
        style={
          Object {
            "height": 967.5,
            "justifyContent": "space-between",
            "maxHeight": 360,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../assets/images/logo.png",
            }
          }
          style={
            Array [
              Object {
                "alignSelf": "center",
                "height": 118,
                "resizeMode": "contain",
                "width": "80%",
              },
              undefined,
            ]
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#F5F5F5",
              "borderColor": "rgba(229, 229, 234, 255)",
              "borderRadius": 15,
              "borderWidth": 1,
              "minHeight": 48,
              "overflow": "hidden",
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "paddingLeft": 16,
                },
                undefined,
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              blurOnSubmit={false}
              keyboardType="email-address"
              maxLength={254}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              onSubmitEditing={[Function]}
              placeholder="Email"
              placeholderTextColor="rgba(174, 174, 178, 255)"
              rejectResponderTermination={true}
              returnKeyType="next"
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "color": "rgba(0,0,0,0.87)",
                    "flex": 1,
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "height": "100%",
                    "letterSpacing": -0.32,
                    "lineHeight": 21,
                  },
                  undefined,
                  undefined,
                ]
              }
              underlineColorAndroid="transparent"
              value={null}
            />
          </View>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "#F5F5F5",
              "borderColor": "rgba(229, 229, 234, 255)",
              "borderRadius": 15,
              "borderWidth": 1,
              "minHeight": 48,
              "overflow": "hidden",
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "paddingLeft": 16,
                },
                undefined,
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              blurOnSubmit={false}
              maxLength={128}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder="Password"
              placeholderTextColor="rgba(174, 174, 178, 255)"
              rejectResponderTermination={true}
              returnKeyType="done"
              secureTextEntry={true}
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "color": "rgba(0,0,0,0.87)",
                    "flex": 1,
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "height": "100%",
                    "letterSpacing": -0.32,
                    "lineHeight": 21,
                  },
                  undefined,
                  Object {
                    "paddingRight": 40,
                  },
                ]
              }
              underlineColorAndroid="transparent"
              value={null}
            />
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "height": 48,
                  "justifyContent": "center",
                  "opacity": 1,
                  "position": "absolute",
                  "right": 0,
                  "width": 48,
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Object {
                    "color": "rgba(174, 174, 178, 255)",
                    "fontFamily": "Ionicons",
                    "fontSize": 30,
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  }
                }
                text="ionicon-ios-eye"
              />
            </View>
          </View>
        </View>
        <View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "rgba(0, 122, 255, 255)",
                "borderRadius": 15,
                "color": "rgba(255, 255, 255, 255)",
                "elevation": 2,
                "flexDirection": "row",
                "height": 48,
                "justifyContent": "center",
                "opacity": 1,
                "shadowColor": "rgba(0, 0, 0, 255)",
                "shadowOffset": Object {
                  "x": 0,
                  "y": 1,
                },
                "shadowOpacity": 0.2,
                "shadowRadius": 1.41,
                "width": "100%",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#FFFFFF",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "letterSpacing": -0.32,
                    "lineHeight": 21,
                  },
                  Array [
                    Object {
                      "color": "rgba(255, 255, 255, 255)",
                    },
                    Object {
                      "backgroundColor": "transparent",
                      "fontFamily": "System",
                      "fontWeight": "600",
                    },
                    undefined,
                  ],
                ]
              }
            >
              SIGN IN
            </Text>
            <ActivityIndicator
              animating={false}
              color="rgba(0, 122, 255, 255)"
              hidesWhenStopped={true}
              size="small"
              style={
                Object {
                  "position": "absolute",
                  "right": 16,
                }
              }
            />
          </View>
          <Text
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": "400",
                  "letterSpacing": -0.32,
                  "lineHeight": 21,
                },
                Array [
                  Object {
                    "color": "rgba(0, 122, 255, 255)",
                  },
                  Object {
                    "backgroundColor": "transparent",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "marginTop": 8,
                    "textAlign": "right",
                  },
                ],
              ]
            }
          >
            Forgot your password?
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          style={
            Object {
              "height": "100%",
              "justifyContent": "space-between",
            }
          }
        >
          <View
            style={
              Object {
                "flex": 3,
                "justifyContent": "space-evenly",
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  Object {
                    "borderColor": "rgba(229, 229, 234, 255)",
                    "borderWidth": 0.5,
                    "flex": 1,
                    "height": 0.5,
                  }
                }
              />
              <Text
                style={
                  Array [
                    Object {
                      "backgroundColor": "transparent",
                      "color": "rgba(0,0,0,0.87)",
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "600",
                      "letterSpacing": -0.154,
                      "lineHeight": 24,
                    },
                    Array [
                      Object {
                        "color": "rgba(0,0,0,0.87)",
                      },
                      Object {
                        "backgroundColor": "transparent",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "marginHorizontal": 8,
                      },
                    ],
                  ]
                }
              >
                Or sign in with
              </Text>
              <View
                style={
                  Object {
                    "borderColor": "rgba(229, 229, 234, 255)",
                    "borderWidth": 0.5,
                    "flex": 1,
                    "height": 0.5,
                  }
                }
              />
            </View>
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "justifyContent": "center",
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "backgroundColor": "rgba(255, 255, 255, 255)",
                    "borderRadius": 10,
                    "opacity": 1,
                  }
                }
                testID="facebook-signin"
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Object {
                      "color": "#3b5998",
                      "fontFamily": "icomoon",
                      "fontSize": 33,
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    }
                  }
                  text="kebeticon-facebook"
                />
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="google-signin"
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../assets/images/google.png",
                    }
                  }
                  style={
                    Object {
                      "height": 33,
                      "marginHorizontal": 8,
                      "width": 32,
                    }
                  }
                />
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "alignItems": "flex-end",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "letterSpacing": -0.32,
                    "lineHeight": 21,
                  },
                  Array [
                    Object {
                      "color": "rgba(0,0,0,0.87)",
                    },
                    Object {
                      "backgroundColor": "transparent",
                      "fontFamily": "System",
                      "fontWeight": "300",
                    },
                    undefined,
                  ],
                ]
              }
            >
              Don't have an account?
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "letterSpacing": -0.32,
                    "lineHeight": 21,
                  },
                  Array [
                    Object {
                      "color": "rgba(0,0,0,0.87)",
                    },
                    Object {
                      "backgroundColor": "transparent",
                      "fontFamily": "System",
                      "fontWeight": "300",
                    },
                    undefined,
                  ],
                ]
              }
            >
               
            </Text>
            <Text
              onPress={[Function]}
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "letterSpacing": -0.32,
                    "lineHeight": 21,
                  },
                  Array [
                    Object {
                      "color": "rgba(0, 122, 255, 255)",
                    },
                    Object {
                      "backgroundColor": "transparent",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    undefined,
                  ],
                ]
              }
            >
              Sign up
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
